### DISCUSSION TITLE
conf.LOGIN_DATA_STORE_PATH.value = None

### DISCUSSION THREAD

**Discussion by Xornop** [femueller/python-n26#101] [Q&A] [answered]
In the examples, you don't go into detail about this option.
I don't know codes that well, suffice to say I am a total noob.
I can edit the `__init__.py`, but to what can i change this so that it actually saves my login data? I have tried replacing `None` with `/config/n26/file.yml` but that doesn't work. What should I change it to? Or how do I go about this?

**Comment by markusressel** [comment]
To set a configuration (including the setting your title suggests) you have to create a configuration **file**, like described in the [Configuration section of the README](https://github.com/femueller/python-n26#configuration).

**Reply by Xornop** [reply]
since the init.py says "None" you can't change the setting in a configuration file, right?
so i wanted to change that "None" in the init file, to something like: 
                conf.LOGIN_DATA_STORE_PATH.value = acc[LOGIN_DATA_STORE_PATH]

and then adding a line withing the config schema:
                    vol.Optional(LOGIN_DATA_STORE_PATH): cv.string,

so I hopefully would be able to provide a path in my configuration file (configuration.yaml) like:
   login_data_store_path: /config/n26

But my python knowledge is not enough to actually get it to work.

**Comment by markusressel** [comment, marked as answer]
> since the init.py says "None" you can't change the setting in a configuration file, right?

I am not sure what `__init__.py` you refer to, there is no such file in the project that contains anything about the `LOGIN_DATA_STORE_PATH` configuration option. But answering the question: yes you can change the setting in the file if you use the CLI or the Api without explicitly and manually creating a `Config` object instance and passing it to the `Api` instance. There is no hardcoded "None" value for this setting, its just the default. But it is only a default, you can of course change it to something else. Otherwise it shouldn't even be a setting afterall.

The configuration file that should be used for using python-n26 is parsed by a library called [container-app-conf](https://github.com/markusressel/container-app-conf). The path (actually, there are multiple) to the configuration file is hardcoded and can be seen [here](https://github.com/markusressel/container-app-conf/blob/579e439fe20d8b665c7587b8adbb087b76046d04/container_app_conf/const.py#L25). Of course this configuration file (yaml or toml file) can **not** contain python code, but only static things like a text (string) or a number.

> and then adding a line withing the config schema:
> vol.Optional(LOGIN_DATA_STORE_PATH): cv.string,

container-app-conf does not need Voluptuous for a simple string config entry. I don't know why or where you would want to add this line.

If you need/want to specify the path to the login data store file programmatically, you have to create an instance of the `Config` yourself and set the configuration options like you mentioned using f.ex.:
```python
conf = Config(validate=False)
conf.LOGIN_DATA_STORE_PATH.value = acc[LOGIN_DATA_STORE_PATH]
conf.validate()
```

However, when using this approach you also need to interact with the `Api` directly and cannot (at least not as easily) use the CLI interface. This is [documented here](https://github.com/femueller/python-n26#api-example).

> so I hopefully would be able to provide a path in my configuration file (configuration.yaml) like:
> login_data_store_path: /config/n26

What prevents you from doing that currently? Both when using the CLI **and** when using the Api in a custom python script, by default, the configuration file is read from the same places. There should not be a reason why doing what you wrote above does not work.

In your initial message you said that you tried to use `/config/n26/file.yml` as a file path. Note that this is an absolute file path on your system! I don't know which linux variant you are running, but using `/config` only for N26 data store doesn't seem right. Also note that the data store is not a yaml file!

Just as an example: I use this configuration file:

```yaml
n26:
  username: myusername@mail.com
  password: "abcdefg"
  login_data_store_path: "~/.config/n26/token_data"
  mfa_type: app
  device_token: some-random-uuid
```

which is located in `~/.config/n26.yml`.
On my system this would resolve to `/home/markus/.config/n26.yml` and  `/home/markus/.config/n26/token_data` respectively.

**Reply by Xornop** [reply]
Hi Markus, thank you for being this helpful.
Should've mentioned, I am using the API via homeassistant (I think?)
I have undone all changes, and changed the line in `__init__.py` from `None` to `conf.LOGIN_DATA_STORE_PATH.value = [~/]`

It now works! Thank you so much!

**Reply by markusressel** [reply]
> Should've mentioned, I am using the API via homeassistant (I think?)

Ahh this changes everything! :smile: 

Yes indeed the integration is outdated. Fyi I am the maintainer of the custom component. The main reason that it hasn't been updated in a long time is the requirement from N26 to re-authenticate through the phone after some time (see #67), which basically makes it unusable. I already thought about removing it altogether, since it would be easy to bring it back if it starts working again.

> `conf.LOGIN_DATA_STORE_PATH.value = [~/]`

I don't quite understand how this is valid python syntax, but if it works for you :+1: :smile:

**Reply by Xornop** [reply]
Me neither, but then again i don't really understand python at all haha
Anyway it now generates a folder called `['~` and within that folder is a file called `']` , containing the login keys so i guess something still goes wrong with the data store file path... what ís correct to get a file called `key` in a subfolder `n26` ?

I thought using the login data store path feature would make n26 remember your login token (or rather the other way around, it would give home assistant a permanent token). And since the integration has "none" as the path, it would not generate a token file. That's the whole reason I am chasing this down lol

**Reply by markusressel** [reply]
> Me neither, but then again i don't really understand python at all haha
> Anyway it now generates a folder called `['~` and within that folder is a file called `']` , containing the login keys so i guess something still goes wrong with the data store file path... what ís correct to get a file called `key` in a subfolder `n26` ?

That doesn't look good xD

> I thought using the login data store path feature would make n26 remember your login token (or rather the other way around, it would give home assistant a permanent token). And since the integration has "none" as the path, it would not generate a token file. That's the whole reason I am chasing this down lol

Well yes and no... The login data store file is indeed used for storing the access token to be able to use it again at a later point in time. However, this still only works for a couple hours. After that a new 2FA approval is required to get a new token. This sucks a** and I would be more than happy if you would let N26 know about this and that we need permanent access tokens.

The "login data store file" feature was originally introduced for a completely different reason: Not having to 2FA authenticate for each and every CLI command call. Home Assistant actually doesn't need to store the login information in a file, since the Api instance is created once and only destroyed when home assistant stops, so it can store the token in memory for the whole duration. But even then the stupid reauth thingy from N26 will kick in at some point, making it impossible to run it in the background for longer periods of time.

**Reply by Xornop** [reply]
I am currently chatting with the support chat, but I don't think 'Vanessa' knows the difference between 'app' and 'API'

**Reply by Xornop** [reply]
You know, sometimes you reboot HA multiple times in an hour, and I just want to not have to login each time I think I am done rebooting it. That's why I needed the file

**Reply by markusressel** [reply]
I would still advise against using the component this way, since it will give you an authentication request notification on your phone every 2 hours, and you can never be sure if it is a legitimate request, since you would have to also check whether home assistant is requesting or some shady middleman that somehow got access to your password. I know its sad, I would really like to use it in home assistant too (I have written the integration...), but its just not feasible with the current authentication system.

**Reply by Xornop** [reply]
I see what you mean. Woke up this morning with 5 authentication notifications.
